src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingContext.cs

```
@@ -145,46 +153,48 @@ internal string CreateCacheKey(CachedVaryBy varyBy)
             {
-                return string.Empty;
-            }
-
-            var builder = new StringBuilder(";");
+                // TODO: resolve key format and delimiters
+                foreach (var header in varyBy.Headers)
+                {
+                    // TODO: Normalization of order, case?
+                    var value = HttpContext.Request.Headers[header];
 
-            foreach (var header in varyBy.Headers)
-            {
-                // TODO: Normalization of order, case?
-                var value = HttpContext.Request.Headers[header].ToString();
+                    // TODO: How to handle null/empty string?
+                    if (StringValues.IsNullOrEmpty(value))
+                    {
+                        value = "null";
+                    }
 
-                // TODO: How to handle null/empty string?
-                if (string.IsNullOrEmpty(value))
-                {
-                    value = "null";
+                    builder.Append(";")
+                        .Append(header)
+                        .Append("=")
+                        .Append(value);
                 }
-
-                builder.Append(header)
-                    .Append("=")
-                    .Append(value)
-                    .Append(";");
             }
+            // TODO: Parse querystring params
 
-            // Parse querystring params
+            // Append custom cache key segment
+            if (Options.CreateCustomizedCacheKey != null)
+            {
+                builder.Append(";")
+                    .Append(Options.CreateCustomizedCacheKey(HttpContext));
+            }
 
             return builder.ToString();
         }
 
         internal bool RequestIsCacheable()
         {
+            // Use optional override if specified by user
+            if (Options.RequestIsCacheableOverride != null)
+            {
+                return Options.RequestIsCacheableOverride(HttpContext);
```

Hmm... should we swallow user exceptions like that? Why not just let it fail the request?